diff --git a/kernel/Kbuild b/kernel/Kbuild
index 889d030..97ba731 100644
--- a/kernel/Kbuild
+++ b/kernel/Kbuild
@@ -148,3 +148,13 @@ ccflags-y += -Wno-strict-prototypes -Wno-int-conversion -Wno-gcc-compat -Wno-mis
 ccflags-y += -Wno-declaration-after-statement -Wno-unused-function -Wno-unused-variable
 
 # Keep a new line here!! Because someone may append config
+
+## For susfs stuff ##
+ifeq ($(shell test -e $(srctree)/fs/susfs.c; echo $$?),0)
+$(eval SUSFS_VERSION=$(shell cat $(srctree)/include/linux/susfs.h | grep -E '^#define SUSFS_VERSION' | cut -d' ' -f3 | sed 's/"//g'))
+$(info )
+$(info -- SUSFS_VERSION: $(SUSFS_VERSION))
+else
+$(info -- You have not integrated susfs in your kernel yet.)
+$(info -- Read: https://gitlab.com/simonpunk/susfs4ksu)
+endif
diff --git a/kernel/Kconfig b/kernel/Kconfig
index c770166..569b571 100644
--- a/kernel/Kconfig
+++ b/kernel/Kconfig
@@ -36,3 +36,90 @@ config KSU_DISABLE_POLICY
 	  non-root handling will follow the global umount policy only.
 
 endmenu
+
+menu "KernelSU - SUSFS"
+config KSU_SUSFS
+	bool "KernelSU addon - SUSFS"
+	depends on KSU
+	depends on THREAD_INFO_IN_TASK
+	default y
+	help
+	  Patch and Enable SUSFS to kernel with KernelSU.
+
+config KSU_SUSFS_SUS_PATH
+	bool "Enable to hide suspicious path (NOT recommended)"
+	depends on KSU_SUSFS
+	default y
+	help
+	  - Allow hiding the user-defined path and all its sub-paths from various system calls.
+	  - Includes temp fix for the leaks of app path in /sdcard/Android/data directory.
+	  - Effective only on zygote spawned user app process.
+	  - Use with cautious as it may cause performance loss and will be vulnerable to side channel attacks,
+	    just disable this feature if it doesn't work for you or you don't need it at all.
+
+config KSU_SUSFS_SUS_MOUNT
+	bool "Enable to hide suspicious mounts"
+	depends on KSU_SUSFS
+	default y
+	help
+	  - Allow hiding the user-defined mount paths from /proc/self/[mounts|mountinfo|mountstat].
+	  - Effective on all processes for hiding mount entries.
+	  - mnt_id and mnt_group_id of the sus mount will be assigned to a much bigger number to solve the issue of id not being contiguous.
+
+config KSU_SUSFS_SUS_KSTAT
+	bool "Enable to spoof suspicious kstat"
+	depends on KSU_SUSFS
+	default y
+	help
+	  - Allow spoofing the kstat of user-defined file/directory.
+	  - Effective only on zygote spawned user app process.
+
+config KSU_SUSFS_SPOOF_UNAME
+	bool "Enable to spoof uname"
+	depends on KSU_SUSFS
+	default y
+	help
+	  - Allow spoofing the string returned by uname syscall to user-defined string.
+	  - Effective on all processes.
+
+config KSU_SUSFS_ENABLE_LOG
+	bool "Enable logging susfs log to kernel"
+	depends on KSU_SUSFS
+	default y
+	help
+	  - Allow logging susfs log to kernel, uncheck it to completely disable all susfs log.
+
+config KSU_SUSFS_HIDE_KSU_SUSFS_SYMBOLS
+	bool "Enable to automatically hide ksu and susfs symbols from /proc/kallsyms"
+	depends on KSU_SUSFS
+	default y
+	help
+	  - Automatically hide ksu and susfs symbols from '/proc/kallsyms'.
+	  - Effective on all processes.
+
+config KSU_SUSFS_SPOOF_CMDLINE_OR_BOOTCONFIG
+	bool "Enable to spoof /proc/bootconfig (gki) or /proc/cmdline (non-gki)"
+	depends on KSU_SUSFS
+	default y
+	help
+	  - Spoof the output of /proc/bootconfig (gki) or /proc/cmdline (non-gki) with a user-defined file.
+	  - Effective on all processes.
+
+config KSU_SUSFS_OPEN_REDIRECT
+	bool "Enable to redirect a path to be opened with another path (experimental)"
+	depends on KSU_SUSFS
+	default y
+	help
+	  - Allow redirecting a target path to be opened with another user-defined path.
+	  - Effective only on processes with uid < 2000.
+
+config KSU_SUSFS_SUS_MAP
+	bool "Enable to hide suspicious maps"
+	depends on KSU_SUSFS
+	default y
+	help
+	  - Allow hiding suspicious maps from /proc/self/[maps|smaps|smaps_rollup|numa_maps|mem|pagemap].
+	  - Allow hiding suspicious map file from /proc/self/map_files/.
+	  - Effective only on zygote spawned user app process.
+
+endmenu
diff --git a/kernel/core/init.c b/kernel/core/init.c
index c158df8..a7d90dc 100644
--- a/kernel/core/init.c
+++ b/kernel/core/init.c
@@ -1,3 +1,6 @@
 #include <linux/export.h>
 #include <linux/fs.h>
+#ifdef CONFIG_KSU_SUSFS
+#include <linux/susfs.h>
+#endif
 #include <linux/kobject.h>
diff --git a/kernel/feature/kernel_umount.c b/kernel/feature/kernel_umount.c
index dd74120..81fdd48 100644
--- a/kernel/feature/kernel_umount.c
+++ b/kernel/feature/kernel_umount.c
@@ -1,3 +1,6 @@
+#ifdef CONFIG_KSU_SUSFS_SUS_PATH
+extern void susfs_run_sus_path_loop(void);
+#endif
 #include <linux/sched.h>
 #include <linux/slab.h>
 #include <linux/task_work.h>
diff --git a/kernel/hook/setuid_hook.c b/kernel/hook/setuid_hook.c
index 3cccc78..7e38083 100644
--- a/kernel/hook/setuid_hook.c
+++ b/kernel/hook/setuid_hook.c
@@ -1,3 +1,20 @@
+
+#ifdef CONFIG_KSU_SUSFS
+#include <linux/susfs_def.h>
+#include "selinux/selinux.h"
+
+static inline bool is_zygote_isolated_service_uid(uid_t uid)
+{
+    uid %= 100000;
+    return (uid >= 99000 && uid < 100000);
+}
+
+static inline bool is_zygote_normal_app_uid(uid_t uid)
+{
+    uid %= 100000;
+    return (uid >= 10000 && uid < 19999);
+}
+#endif
 #include <linux/compiler.h>
 #include <linux/version.h>
 #include <linux/slab.h>
diff --git a/kernel/runtime/ksud_integration.c b/kernel/runtime/ksud_integration.c
index 38d1476..6c88478 100644
--- a/kernel/runtime/ksud_integration.c
+++ b/kernel/runtime/ksud_integration.c
@@ -1,3 +1,6 @@
+#ifdef CONFIG_KSU_SUSFS
+extern int ksu_handle_execveat_init(const char *path);
+#endif
 #include "feature/selinux_hide.h"
 #include <linux/rcupdate.h>
 #include <linux/slab.h>
diff --git a/kernel/selinux/selinux.c b/kernel/selinux/selinux.c
index 88c7956..ccabc7e 100644
--- a/kernel/selinux/selinux.c
+++ b/kernel/selinux/selinux.c
@@ -226,3 +226,83 @@ void escape_to_root_for_adb_root(void)
     }
     commit_creds(cred);
 }
+
+#ifdef CONFIG_KSU_SUSFS
+#define KERNEL_INIT_DOMAIN "u:r:init:s0"
+#define KERNEL_ZYGOTE_DOMAIN "u:r:zygote:s0"
+#define KERNEL_PRIV_APP_DOMAIN "u:r:priv_app:s0:c512,c768"
+
+u32 susfs_ksu_sid = 0;
+u32 susfs_init_sid = 0;
+u32 susfs_zygote_sid = 0;
+u32 susfs_priv_app_sid = 0;
+
+static inline void susfs_set_sid(const char *secctx_name, u32 *out_sid)
+{
+    int err;
+    if (!secctx_name || !out_sid)
+        return;
+    err = security_secctx_to_secid(secctx_name, strlen(secctx_name), out_sid);
+    if (err)
+        pr_err("susfs: failed to get sid for %s, err: %d\n", secctx_name, err);
+}
+
+void susfs_set_zygote_sid(void)
+{
+    susfs_set_sid(KERNEL_ZYGOTE_DOMAIN, &susfs_zygote_sid);
+}
+
+bool susfs_is_current_zygote_domain(void)
+{
+    return susfs_is_sid_equal(current_cred(), susfs_zygote_sid);
+}
+
+void susfs_set_ksu_sid(void)
+{
+    susfs_set_sid(KERNEL_SU_DOMAIN, &susfs_ksu_sid);
+}
+
+bool susfs_is_current_ksu_domain(void)
+{
+    return susfs_is_sid_equal(current_cred(), susfs_ksu_sid);
+}
+
+void susfs_set_init_sid(void)
+{
+    susfs_set_sid(KERNEL_INIT_DOMAIN, &susfs_init_sid);
+}
+
+bool susfs_is_current_init_domain(void)
+{
+    return susfs_is_sid_equal(current_cred(), susfs_init_sid);
+}
+
+void susfs_set_priv_app_sid(void)
+{
+    susfs_set_sid(KERNEL_PRIV_APP_DOMAIN, &susfs_priv_app_sid);
+}
+
+bool susfs_is_sid_equal(const struct cred *cred, u32 sid2)
+{
+    if (!cred || !sid2)
+        return false;
+    return (cred->security && ((struct task_security_struct *)cred->security)->sid == sid2);
+}
+
+u32 susfs_get_current_sid(void)
+{
+    const struct cred *cred = current_cred();
+    if (!cred || !cred->security)
+        return 0;
+    return ((struct task_security_struct *)cred->security)->sid;
+}
+
+u32 susfs_get_sid_from_name(const char *secctx_name)
+{
+    u32 sid = 0;
+    if (!secctx_name)
+        return 0;
+    security_secctx_to_secid(secctx_name, strlen(secctx_name), &sid);
+    return sid;
+}
+#endif
diff --git a/kernel/selinux/selinux.h b/kernel/selinux/selinux.h
index a17ed22..2b1bb95 100644
--- a/kernel/selinux/selinux.h
+++ b/kernel/selinux/selinux.h
@@ -40,3 +40,16 @@ void escape_to_root_for_adb_root();
 extern u32 ksu_file_sid;
 
 #endif
+
+#ifdef CONFIG_KSU_SUSFS
+bool susfs_is_sid_equal(const struct cred *cred, u32 sid2);
+u32 susfs_get_sid_from_name(const char *secctx_name);
+u32 susfs_get_current_sid(void);
+void susfs_set_zygote_sid(void);
+bool susfs_is_current_zygote_domain(void);
+void susfs_set_ksu_sid(void);
+bool susfs_is_current_ksu_domain(void);
+void susfs_set_init_sid(void);
+bool susfs_is_current_init_domain(void);
+void susfs_set_priv_app_sid(void);
+#endif
diff --git a/kernel/supercall/dispatch.c b/kernel/supercall/dispatch.c
index 902631d..8990b06 100644
--- a/kernel/supercall/dispatch.c
+++ b/kernel/supercall/dispatch.c
@@ -1,3 +1,8 @@
+#ifdef CONFIG_KSU_SUSFS
+#include <linux/namei.h>
+#include <linux/susfs.h>
+#include <linux/susfs_def.h>
+#endif
 #include <linux/capability.h>
 #include <linux/cred.h>
 #include <linux/slab.h>
@@ -992,3 +997,79 @@ void ksu_supercall_cleanup_state(void)
     }
     up_write(&mount_list_lock);
 }
+
+#ifdef CONFIG_KSU_SUSFS
+int ksu_handle_susfs_cmd(unsigned int cmd, void __user **arg)
+{
+    switch (cmd) {
+#ifdef CONFIG_KSU_SUSFS_SUS_PATH
+    case CMD_SUSFS_ADD_SUS_PATH:
+        susfs_add_sus_path(arg);
+        return 0;
+    case CMD_SUSFS_ADD_SUS_PATH_LOOP:
+        susfs_add_sus_path_loop(arg);
+        return 0;
+#endif
+#ifdef CONFIG_KSU_SUSFS_SUS_MOUNT
+    case CMD_SUSFS_HIDE_SUS_MNTS_FOR_NON_SU_PROCS:
+        susfs_set_hide_sus_mnts_for_non_su_procs(arg);
+        return 0;
+#endif
+#ifdef CONFIG_KSU_SUSFS_SUS_KSTAT
+    case CMD_SUSFS_ADD_SUS_KSTAT:
+        susfs_add_sus_kstat(arg);
+        return 0;
+    case CMD_SUSFS_UPDATE_SUS_KSTAT:
+        susfs_update_sus_kstat(arg);
+        return 0;
+    case CMD_SUSFS_ADD_SUS_KSTAT_STATICALLY:
+        susfs_add_sus_kstat(arg);
+        return 0;
+#ifdef CONFIG_KSU_SUSFS_SUS_KSTAT_REDIRECT
+    case CMD_SUSFS_ADD_SUS_KSTAT_REDIRECT:
+        susfs_add_sus_kstat_redirect(arg);
+        return 0;
+#endif
+#endif
+#ifdef CONFIG_KSU_SUSFS_SPOOF_UNAME
+    case CMD_SUSFS_SET_UNAME:
+        susfs_set_uname(arg);
+        return 0;
+#endif
+#ifdef CONFIG_KSU_SUSFS_ENABLE_LOG
+    case CMD_SUSFS_ENABLE_LOG:
+        susfs_enable_log(arg);
+        return 0;
+#endif
+#ifdef CONFIG_KSU_SUSFS_SPOOF_CMDLINE_OR_BOOTCONFIG
+    case CMD_SUSFS_SET_CMDLINE_OR_BOOTCONFIG:
+        susfs_set_cmdline_or_bootconfig(arg);
+        return 0;
+#endif
+#ifdef CONFIG_KSU_SUSFS_OPEN_REDIRECT
+    case CMD_SUSFS_ADD_OPEN_REDIRECT:
+        susfs_add_open_redirect(arg);
+        return 0;
+#endif
+#ifdef CONFIG_KSU_SUSFS_SUS_MAP
+    case CMD_SUSFS_ADD_SUS_MAP:
+        susfs_add_sus_map(arg);
+        return 0;
+#endif
+    case CMD_SUSFS_ENABLE_AVC_LOG_SPOOFING:
+        susfs_set_avc_log_spoofing(arg);
+        return 0;
+    case CMD_SUSFS_SHOW_ENABLED_FEATURES:
+        susfs_get_enabled_features(arg);
+        return 0;
+    case CMD_SUSFS_SHOW_VARIANT:
+        susfs_show_variant(arg);
+        return 0;
+    case CMD_SUSFS_SHOW_VERSION:
+        susfs_show_version(arg);
+        return 0;
+    default:
+        return -EINVAL;
+    }
+}
+#endif
diff --git a/kernel/supercall/supercall.c b/kernel/supercall/supercall.c
index 1a39732..0d72052 100644
--- a/kernel/supercall/supercall.c
+++ b/kernel/supercall/supercall.c
@@ -1,3 +1,9 @@
+#ifdef CONFIG_KSU_SUSFS
+#include <linux/namei.h>
+#include <linux/susfs.h>
+#include <linux/susfs_def.h>
+extern int ksu_handle_susfs_cmd(unsigned int cmd, void __user **arg);
+#endif
 #include <linux/anon_inodes.h>
 #include <linux/err.h>
 #include <linux/fdtable.h>
