diff --git a/kernel/supercall/dispatch.c b/kernel/supercall/dispatch.c
index c1d3aba..bea4290 100644
--- a/kernel/supercall/dispatch.c
+++ b/kernel/supercall/dispatch.c
@@ -422,6 +422,11 @@ int ksu_handle_sys_reboot(int magic1, int magic2, unsigned int cmd, void __user
         case CMD_SUSFS_ADD_SUS_KSTAT_STATICALLY:
             susfs_add_sus_kstat(arg);
             return 0;
+#ifdef CONFIG_KSU_SUSFS_SUS_KSTAT_REDIRECT
+        case CMD_SUSFS_ADD_SUS_KSTAT_REDIRECT:
+            susfs_add_sus_kstat_redirect(arg);
+            return 0;
+#endif
 #endif // #ifdef CONFIG_KSU_SUSFS_SUS_KSTAT
 #ifdef CONFIG_KSU_SUSFS_SPOOF_UNAME
         case CMD_SUSFS_SET_UNAME:
