diff --git a/include/linux/susfs.h b/include/linux/susfs.h
index 2819a07..c72e322 100644
--- a/include/linux/susfs.h
+++ b/include/linux/susfs.h
@@ -1,6 +1,24 @@
 #ifndef KSU_SUSFS_H
 #define KSU_SUSFS_H
 
+struct st_susfs_sus_kstat_redirect {
+	char                                    virtual_pathname[SUSFS_MAX_LEN_PATHNAME];
+	char                                    real_pathname[SUSFS_MAX_LEN_PATHNAME];
+	unsigned long                           spoofed_ino;
+	unsigned long                           spoofed_dev;
+	unsigned int                            spoofed_nlink;
+	long long                               spoofed_size;
+	long                                    spoofed_atime_tv_sec;
+	long                                    spoofed_mtime_tv_sec;
+	long                                    spoofed_ctime_tv_sec;
+	long                                    spoofed_atime_tv_nsec;
+	long                                    spoofed_mtime_tv_nsec;
+	long                                    spoofed_ctime_tv_nsec;
+	unsigned long                           spoofed_blksize;
+	unsigned long long                      spoofed_blocks;
+	int                                     err;
+};
+
 #include <linux/version.h>
 #include <linux/types.h>
 #include <linux/utsname.h>
@@ -238,4 +256,11 @@ void susfs_start_sdcard_monitor_fn(void);
 /* susfs_init */
 void susfs_init(void);
 
+#ifdef CONFIG_KSU_SUSFS_SUS_KSTAT_REDIRECT
+int susfs_add_sus_kstat_redirect(void __user *user_arg);
+bool susfs_is_sus_kstat_redirect(struct dentry *dentry, struct kstat *stat);
+#endif
+#ifdef CONFIG_KSU_SUSFS_UNICODE_FILTER
+bool susfs_check_unicode_bypass(const char *pathname);
+#endif
 #endif
\ No newline at end of file
